[CKC] Add CVE verification with --check-cves option

bmastbergen opened the pull request:
Adds ability to verify that CVE references in PR commit messages correctly match the upstream commits they reference. Uses the kernel vulnerabilities database to cross-check CVE assignments against upstream commit hashes.

The --check-cves flag enables validation that detects three error conditions: mismatched CVE assignments between PR and upstream commits, CVE references to upstream commits with no CVE assignment, and failures accessing the vulnerabilities database. Output format matches existing checker patterns with support for both plain text and markdown modes.

Here is an example github comment from this code:
https://github.com/ctrliq/kernel-src-tree/pull/616#issuecomment-3386172159

This was written by Claude BTW.

bmastbergen committed:
[CKC] Add CVE verification with --check-cves option

Adds ability to verify that CVE references in PR commit messages correctly
match the upstream commits they reference. Uses the kernel vulnerabilities
database to cross-check CVE assignments against upstream commit hashes.

The --check-cves flag enables validation that detects three error conditions:
mismatched CVE assignments between PR and upstream commits, CVE references
to upstream commits with no CVE assignment, and failures accessing the
vulnerabilities database. Output format matches existing checker patterns
with support for both plain text and markdown modes.

ciq-it-service-account commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/ciq-default/project/5bb6dc7b-bdbf-4ac0-8cb1-18ccae130bd8/pr-check/e13db1a0-1417-4a3b-85f3-75e32108a969?refs=da974d35f96714635e63dc313e373a45897070e0&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

bmastbergen commented on check_kernel_commits.py:
The future is now copilot!  Its 2025!

bmastbergen commented on check_kernel_commits.py:
Probably fine the way it is though

PlaidCat commented on check_kernel_commits.py:
Oh Co-Pilot, your contexted didn't even follow the comment directly above which you also whined about.

PlaidCat commented on check_kernel_commits.py:
```diff
@@ -185,10 +271,97 @@ def main():
                         out_lines.append('    ' + line)
                     out_lines.append("")  # blank line
 
+            # Check CVE if enabled
```
I really hate this tree of if statements.

Can we make a ticket to clean this up ... ideally as a part of python mock testing ?

roxanan1996 replied:
@PlaidCat Yees, I fully agree with you! On my list to refactor this when I migrate to 'kt'

roxanan1996 commented on check_kernel_commits.py:
no need to use str(e), just e is ok. Same below.

bmastbergen replied:
Fixed.  Thanks!

roxanan1996 commented on check_kernel_commits.py:
I would use Optiona[str] here as result

bmastbergen replied:
Added.  Thanks!

roxanan1996 replied:
I would not return the bool, just the string being not None should be enough, but it's fine for now. Could be improved later.

PlaidCat reviewed (dismissed):
This is too useful to block on a tree of if's

roxanan1996 commented on check_kernel_commits.py:
```diff
@@ -168,8 +228,34 @@ def main():
+                                    fix_cves[fix_hash] = bugfix_cve
+                        except (RuntimeError, subprocess.SubprocessError):
+                            # Log a warning instead of silently ignoring errors when checking bugfix CVEs
+                            print(f"Warning: Failed to check CVE for bugfix commit {fix_hash}: {e}", file=sys.stderr)
```
e is undefined

bmastbergen replied:
Fixed.  Thanks!

roxanan1996 reviewed (commented):
This script needs refactoring soon, as it becoming bigger and bigger but for now, it's fine. 

One question: this does not cover the case where there are multiple commits upstream for a CVE but the pull requests does not include all of them, right?

PlaidCat commented:
> This script needs refactoring soon, as it becoming bigger and bigger but for now, it's fine.
> 
> One question: this does not cover the case where there are multiple commits upstream for a CVE but the pull requests does not include all of them, right?

Do you have an example of this since Kernel.org became the CNA.
I believe this is an artifact of the past when kernel.org had no control over CVEs cut against it.

it might be useful to some pre 2023 but we have no way of validating those within the `kernel.org/vulns` project (they're a bit of a mess)

bmastbergen commented:
Addressed comments from @roxanan1996 

I agree with both of you that this script could use some refactoring.  Lets get it in and checking PRs then fix it up.  Thanks!
